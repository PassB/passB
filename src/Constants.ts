export enum executionContext {
  background = 'background',
  popup = 'popup',
  options = 'options',
}
