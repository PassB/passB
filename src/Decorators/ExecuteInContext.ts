// tslint:disable:no-any

import 'reflect-metadata';

export type ExecutionContext = 'background' | 'popup' | 'options' | undefined;
const executionContexts = ['background', 'popup'];

function getActionName(constructor: Function, propertyKey: string): string {
  return `${constructor.name}.${propertyKey}`;
}

interface AsynchronousRequest {
  action: string;
  params: any[];
}

export const executeInCorrectContext = () =>
  <R, T extends (this: any, ...args: any[]) => Promise<R>>
    (
      target: object,
      propertyKey: string,
      descriptor: TypedPropertyDescriptor<T>,
    ): TypedPropertyDescriptor<T> => {

    const executionContext = Reflect.getMetadata('executionContext', target, propertyKey);
    if (!executionContexts.includes(executionContext)) {
      throw new Error(`using ${executeInCorrectContext.name} `
        + 'decorator, but not declaring @Reflect.metadata("executionContext", "...") correctly',
      );
    }

    const wrappedFunction = descriptor.value || (() => Promise.reject(void 0));

    return {
      value: (async function wrapper(this: any, ...params: any[]): Promise<R> {
        if (window.executionContext === executionContext) {
          // console.debug('executing synchonously %s.%s', target.constructor.name, propertyKey);
          return await wrappedFunction.apply(this, params);
        }

        // console.debug('executing asynchonously %s.%s', target.constructor.name, propertyKey);
        return await browser.runtime.sendMessage({
          action: getActionName(target.constructor, propertyKey),
          params: [...params],
        });
      }) as T,
    };
  };

export const AsynchronousCallable = () =>
  <T extends { new(...args: any[]): {} }>(constructor: T): T =>
    class extends constructor {
      public constructor(...args: any[]) {
        super(...args);

        for (let obj = this; !!obj; obj = Object.getPrototypeOf(obj)) { // tslint:disable-line:no-this-assignment
          for (const propertyKey of Object.getOwnPropertyNames(obj)) {
            const executionContext = Reflect.getMetadata('executionContext', this, propertyKey);
            if (executionContext && window.executionContext === executionContext) {

              browser.runtime.onMessage.addListener(
                (request: AsynchronousRequest, sender: object, sendResponse: Function) => {
                  if (request.action !== getActionName(constructor, propertyKey)) {
                    return false;
                  }

                  const wrappedFn: ((...args: any[]) => Promise<any>) = (this as any)[propertyKey];
                  wrappedFn.apply(this, request.params).then(sendResponse);
                  return true;
                },
              );

            }
          }
        }
      }
    };

export const AsynchronousCallableServiceFactory =
  <X, T extends { new(...args: any[]): X }>(type: T): ((...args: any[]) => X) =>
    ((...args: any[]): X => (new (AsynchronousCallable()(type))(...args)));
