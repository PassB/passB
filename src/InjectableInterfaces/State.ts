import {Store} from 'redux';
import {Persistor} from 'redux-persist';
import {StoreContents} from '../State/State';

export interface State {
  hydrated: Promise<void>;

  getPersistor(): Persistor;

  getStore(): Store<StoreContents>;
}
