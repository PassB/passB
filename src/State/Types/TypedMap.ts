import {fromJS, List, Map} from 'immutable';

type SimpleValues = string | number | boolean | RegExp;

type AllowedValue =
  SimpleValues |
  AllowedMap |
  AllowedList |
  TypedMap<any> | // tslint:disable-line:no-any
  undefined;

interface AllowedList extends List<AllowedValue> {
}

interface AllowedMap extends Map<string, AllowedValue> {
}

export type MapTypeAllowedData<DataType> = {
  [K in keyof DataType]: AllowedValue;
};

export interface TypedMap<DataType extends MapTypeAllowedData<DataType>> extends Map<string, AllowedValue> { // tslint:disable-line:no-any
  toJS(): DataType;

  get<K extends keyof DataType>(key: K, notSetValue?: DataType[K]): DataType[K];

  set<K extends keyof DataType>(key: K, value: DataType[K]): this;
}

export function createTypedMap<DataType extends MapTypeAllowedData<DataType>>(data: DataType): TypedMap<DataType> {
  return fromJS(data);
}
