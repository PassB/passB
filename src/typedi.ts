// this file is here only temporarily until typedi has merged some changes
// and we can use the upstream typedi version instead of a github commit

export * from 'typedi/build/compiled/src';
