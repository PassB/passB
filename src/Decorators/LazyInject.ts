import {Container, Token} from "typedi";

export function LazyInject(typeOrName?: ((type?: any) => Function) | string | Token<any>): Function {
  return (target: Object, propertyName: string, index?: number) => {

    let injectedService: any;

    Object.defineProperty(target, propertyName, {
      get: () => {
        if (!injectedService) {
          if (!typeOrName) {
            typeOrName = () => (Reflect as any).getMetadata("design:type", target, propertyName);
          }

          let identifier: any;
          if (typeof typeOrName === "string") {
            identifier = typeOrName;

          } else if (typeOrName instanceof Token) {
            identifier = typeOrName;

          } else if (typeof typeOrName === 'function') {
            identifier = typeOrName();
          }

          injectedService = Container.get(identifier);
        }
        return injectedService;
      },
      enumerable: true,
      configurable: false,
    });
  };
}
