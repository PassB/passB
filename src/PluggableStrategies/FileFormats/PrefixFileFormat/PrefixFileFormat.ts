import {Service} from 'typedi';
import {OptionsPanelType} from 'Options/OptionsReceiver';
import {createTypedMap} from 'State/Types/TypedMap';
import {FileFormat, FileFormatTag} from '../';
import {OptionsPanel} from './OptionsPanel';

export interface Options {
  passwordFirstLine: boolean;
  passwordPrefix: string;
  usernamePrefix: string;
  trimWhitespace: boolean;
}

@Service({tags: [FileFormatTag]})
export class PrefixFileFormat extends FileFormat<Options> {
  public readonly OptionsPanel: OptionsPanelType<Options> = OptionsPanel;

  public constructor() {
    super(
      'PrefixFileFormat',
      createTypedMap({
        passwordFirstLine: true,
        passwordPrefix: '',
        usernamePrefix: 'login:',
        trimWhitespace: true,
      }),
    );
  }

  public getPassword(lines: string[], entryName: string): string | undefined {
    if (this.options.get('passwordFirstLine')) {
      return lines[0];
    }
    return this.getEntryByPrefix(lines, this.options.get('passwordPrefix'));
  }

  public getUsername(lines: string[], entryName: string): string | undefined {
    return this.getEntryByPrefix(lines, this.options.get('usernamePrefix'));
  }

  private getEntryByPrefix(lines: string[], prefix: string): string | undefined {
    for (const line of lines) {
      if (line.startsWith(prefix)) {
        let entry = line.substr(prefix.length);
        if (this.options.get('trimWhitespace')) {
          entry = entry.trim();
        }
        return entry;
      }
    }
    return;
  }
}
