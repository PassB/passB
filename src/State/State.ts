import {fromJS, List, Map} from 'immutable';
import {
  applyMiddleware, combineReducers, createStore, Action, Reducer, Store,
} from 'redux';
import {persistReducer, persistStore, Persistor} from 'redux-persist';
import immutableTransform = require('redux-persist-transform-immutable');
import { composeWithDevTools } from 'remote-redux-devtools';
import {Service} from 'typedi';
import {LazyInject} from '../Decorators/LazyInject';
import {storageSyncListener, storageSyncMiddleware} from './syncMiddleware';
import {BrowserStorageAdapter, StorageAdaper} from './BrowserStorageAdapter';
import {reducer as OptionsReducer, OptionsState} from './Options';

export interface StoreContents {
  options: OptionsState;
}

export const getOptionsFromState = (state: StoreContents) => state.options;

@Service()
export class State {
  public readonly hydrated: Promise<void>;
  private store: Store<StoreContents>;
  private persistor: Persistor;

  @LazyInject(() => BrowserStorageAdapter)
  private storageAdapter: StorageAdaper;

  public constructor() {
    const reducer: Reducer<StoreContents> = persistReducer(
      {
        key: 'root',
        storage: this.storageAdapter,
        transforms: [immutableTransform({})],
      },
      combineReducers({
        options: OptionsReducer,
      }),
    );

    this.store = createStore(
      reducer,
      composeWithDevTools(
        applyMiddleware(
          storageSyncMiddleware(this.storageAdapter),
        ),
      ),
    );

    this.persistor = persistStore(this.store);
    this.hydrated = new Promise((done: () => void) => {
      let unsubscribe: () => void;
      const checkHydrated = () => {
        if (this.persistor.getState()) {
          if (unsubscribe) {
            unsubscribe();
          }
          done();
        }
      };
      unsubscribe = this.persistor.subscribe(checkHydrated);
      checkHydrated();
    });

    storageSyncListener(this.store, this.storageAdapter);
  }

  public getOptions(): OptionsState {
    return getOptionsFromState(this.store.getState());
  }

  public getPersistor(): Persistor {
    return this.persistor;
  }

  public getStore(): Store<StoreContents> {
    return this.store;
  }

  public dispatch(action: Action): void {
    this.hydrated.then(() => this.store.dispatch(action));
  }
}
