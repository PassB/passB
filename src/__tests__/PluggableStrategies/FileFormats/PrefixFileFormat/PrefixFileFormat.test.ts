import {Container} from 'typedi';
import {executionContext} from 'Constants';
import {setExecutionContext} from 'Decorators/ExecuteInContext';
import {Options, PrefixFileFormat} from 'PluggableStrategies/FileFormats/PrefixFileFormat/PrefixFileFormat';
import {State} from 'State/State';
import {createTypedMap, TypedMap} from 'State/Types/TypedMap';

const passwordFileName1 = 'test/file/path';

const passwordFile1 = [
  'firstLinePw',
  'username: testUser',
  'login: testUser2',
  'password: specifiedPw',
];

const getOptions = (partial: Partial<Options> = {}): TypedMap<Options> => createTypedMap({
  passwordFirstLine: true,
  passwordPrefix: '',
  usernamePrefix: 'login:',
  trimWhitespace: true,
  ...partial,
});

describe('PrefixFileFormat', () => {

  const optionsGetter = jest.fn(getOptions);

  beforeEach(() => {
    optionsGetter.mockReset();
    Container.remove(PrefixFileFormat);
    Object.defineProperty(
      Container.get(PrefixFileFormat),
      'options',
      {get: optionsGetter},
    );
  });

  it('dispatches expected defaultOptions in background context', () => {
    const dispatch = jest.fn();

    setExecutionContext(executionContext.background);
    Container.set(State, {dispatch});
    Container.remove(PrefixFileFormat);
    Container.get(PrefixFileFormat);

    expect(dispatch).toHaveBeenCalled();
    expect(dispatch.mock.calls).toMatchSnapshot();

    Container.remove(State);
  });

  it('should return first line when passwordFirstLine is true', () => {
    optionsGetter.mockReturnValue(getOptions({
      passwordFirstLine: true,
    }));

    expect(Container.get(PrefixFileFormat).getPassword(passwordFile1, passwordFileName1)).toBe('firstLinePw');
  });

  it('should return password when passwordPrefix is found', () => {
    optionsGetter.mockReturnValue(getOptions({
      passwordFirstLine: false,
      passwordPrefix: 'password:',
    }));

    expect(Container.get(PrefixFileFormat).getPassword(passwordFile1, passwordFileName1)).toBe('specifiedPw');
  });

  it('should return undefined when passwordPrefix is not present', () => {
    optionsGetter.mockReturnValue(getOptions({
      passwordFirstLine: false,
      passwordPrefix: 'password_not_present:',
    }));

    expect(Container.get(PrefixFileFormat).getPassword(passwordFile1, passwordFileName1)).toBeUndefined();
  });

  it('should return username when usernamePrefix is found', () => {
    optionsGetter.mockReturnValue(getOptions({
      usernamePrefix: 'username:',
    }));

    expect(Container.get(PrefixFileFormat).getUsername(passwordFile1, passwordFileName1)).toBe('testUser');
  });

  it('should return undefined when usernamePrefix is not present', () => {
    optionsGetter.mockReturnValue(getOptions({
      usernamePrefix: 'username_not_present:',
    }));

    expect(Container.get(PrefixFileFormat).getUsername(passwordFile1, passwordFileName1)).toBeUndefined();
  });

  it('should not trim matches when trimWhitespace is set to false', () => {
    optionsGetter.mockReturnValue(getOptions({
      passwordFirstLine: false,
      usernamePrefix: 'username:',
      trimWhitespace: false,
      passwordPrefix: 'password:',
    }));

    const fileFormat = Container.get(PrefixFileFormat);
    expect(fileFormat.getPassword(passwordFile1, passwordFileName1)).toBe(' specifiedPw');
    expect(fileFormat.getUsername(passwordFile1, passwordFileName1)).toBe(' testUser');
  });
});
