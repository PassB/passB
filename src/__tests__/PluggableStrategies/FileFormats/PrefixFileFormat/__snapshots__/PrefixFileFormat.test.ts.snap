// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrefixFileFormat dispatches expected defaultOptions in background context 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "options": Immutable.Map {
          "passwordFirstLine": true,
          "passwordPrefix": "",
          "usernamePrefix": "login:",
          "trimWhitespace": true,
        },
        "strategyName": "PrefixFileFormat",
        "strategyType": "FileFormat",
      },
      "type": "OPTIONS/SET_STRATEGY_DEFAULT_OPTIONS",
    },
  ],
]
`;
